sql/analyzer/index_dyn.go

```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright (c) 2000, 2099, ducesoft and/or its affiliates. All rights reserved.
+ * DUCESOFT PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
+ *
+ *
+ */
+
+package analyzer
+
+import "github.com/dolthub/go-mysql-server/sql"
+
+const DynIndex = "DYN"
+
+type DynamicIndex interface {
```

Comment:
I don't think you want or need a new interface to accomplish your goals here. If you want to implement a custom index type that matches some subset of expressions, your best bet is for your table implementation to return N indexes, one for each combination of expressions you are capable of indexing. This is how Dolt originally implemented index prefix matching before we implemented it natively in go-mysql-server.

Reply:
It's difficult to match any index without query condition,  'IndexAddressable' not support query condition in go-mysql-server now. This solution is the simplest way to implement some index feature without code changes. Virtual table need dynamic index to skip scan all table.

Reply:
If use'IndexAddressable' for table implementation, it need build too many complex index to match query. The number of columns combinations is very big and complex.

Reply:
One thing we could maybe support is a new interface that does something like this:
```go
type IndexSearchable interface {
    IndexWithPrefix([]string) []sql.Index
}
```

You would switch for this type before falling back to the old expression checking pattern.

We are going to delete indexAnalyzer and all of the existing index matching code shortly, to be replaced with a very different way of representing indexes and expressions. New index matching interfaces needs to be compatible with how this logic will work in a couple months

Reply:
Max's suggestion sounds reasonable. The rest of the PR looks reasonable if you make that change.

Thanks for the contribution!